File: Assets/Scripts/Frame/Thread/CustomThread.cs
using System;
using System.Collections.Generic;
using System.Threading;

public class CustomThread : GameBase
{
	protected CustomThreadCallback mCallback;
	protected ThreadTimeLock mTimeLock;
	protected Thread mThread;
	protected string mName;
	protected bool mRunning;
	protected bool mFinish;
	protected bool mPause;
	protected bool mIsBackground;       // 是否为后台线程,如果是后台线程,则在应用程序关闭时,子线程会自动强制关闭
	public CustomThread(string name)
	{
		mName = name;
		mCallback = null;
		mThread = null;
		mTimeLock = null;
		mFinish = true;
		mRunning = false;
		mPause = false;
		mIsBackground = true;
	}
	public void destroy()
	{
		stop();
	}
	public void setBackground(bool background)
	{
		mIsBackground = background;
		if (mThread != null)
		{
			mThread.IsBackground = mIsBackground;
		}
	}
	public void start(CustomThreadCallback callback, int frameTimeMS = 15, int forceSleep = 5)
	{
		if (mThread != null)
		{
			return;
		}
		mTimeLock = new ThreadTimeLock(frameTimeMS);
		mTimeLock.setForceSleep(forceSleep);
		mRunning = true;
		mCallback = callback;
		mThread = new Thread(run);
		mThread.Name = mName;
		mThread.Start();
		mThread.IsBackground = mIsBackground;
		logInfo("线程启动成功 : " + mName, LOG_LEVEL.LL_FORCE);
	}
	public void pause(bool isPause) { mPause = isPause; }
	public bool isFinished() { return mFinish; }
	public void stop()
	{
		if (mThread == null)
		{
			return;
		}
		mRunning = false;
		while (!mIsBackground && !mFinish) { }
		mThread?.Abort();
		mThread = null;
		mCallback = null;
		mTimeLock = null;
		mPause = false;
		logInfo("线程退出完成! 线程名 : " + mName, LOG_LEVEL.LL_FORCE);
	}
	protected void run()
	{
		mFinish = false;
		while (mRunning)
		{
			mTimeLock.update();
			try
			{
				if (mPause)
				{
					continue;
				}
				bool run = true;
				mCallback(ref run);
				if (!run)
				{
					break;
				}
			}
			catch (Exception e)
			{
				logError("捕获线程异常! 线程名 : " + mName + ", " + e.Message + ", " + e.StackTrace, false);
			}
		}
		mFinish = true;
	}
}

File: ToolProject/AtlasPacker/AtlasPacker/Thread/CustomThread.cpp
#include "CustomThread.h"
#include "txMemoryTrace.h"
#include "TimeLock.h"

CustomThread::CustomThread(const string& name)
{
	mName = name;
	mFinish = true;
	mRunning = false;
	mCallback = NULL;
	mThread = NULL_THREAD;
	mTimeLock = NULL;
	mPause = false;
	mIsBackground = true;
}

CustomThread::~CustomThread()
{
	destroy();
}

void CustomThread::destroy()
{
	stop();
}

void CustomThread::start(CustomThreadCallback callback, void* args, int frameTimeMS)
{
	//LOG_INFO("准备启动线程 : %s", mName.c_str());
	if (mThread != NULL_THREAD)
	{
		//LOG_ERROR("线程已经启动 : %s", mName.c_str());
		return;
	}
	mTimeLock = NEW(TimeLock, mTimeLock, frameTimeMS);
	mRunning = true;
	mCallback = callback;
	mArgs = args;
	CREATE_THREAD(mThread, run, this);
	//LOG_INFO("线程启动成功 : %s", mName.c_str());
}

void CustomThread::stop()
{
	//LOG_INFO("准备退出线程 : %s", mName.c_str());
	mRunning = false;
	mPause = false;
	while (!mIsBackground && !mFinish) {}
	CLOSE_THREAD(mThread);
	mCallback = NULL;
	DELETE(mTimeLock);
	//LOG_INFO("线程退出完成! 线程名 : %s", mName.c_str());
}

void CustomThread::updateThread()
{
	mFinish = false;
	while (mRunning)
	{
		mTimeLock->update();
		try
		{
			if (mPause)
			{
				continue;
			}
			if (!mCallback(mArgs))
			{
				break;
			}
		}
		catch (...)
		{
			//LOG_INFO("捕获线程异常! 线程名 : %s", mName.c_str());
		}
	}
	mFinish = true;
}